docs/development/logging.md

```diff
@@ -0,0 +1,165 @@
+# Logging in Gardener Components
+
+This document aims at providing a general developer guideline on different aspects of logging practices and conventions used in the Gardener codebase.
+It contains mostly Gardener-specific points and references other existing and commonly accepted logging guidelines for general advice.
+Developers and reviewers should consult this guide when writing, refactoring and reviewing Gardener code.
+If parts are unclear or new learnings arise, this guide should be adapted accordingly.
+
+## Logging Libraries / Implementations
+
+Historically, Gardener components have been using [logrus](https://github.com/sirupsen/logrus).
+There is a global logrus logger ([`logger.Logger`](https://github.com/gardener/gardener/blob/626ba7c10e1150819b3905116d3988512c18c9ee/pkg/logger/logrus.go#L28)) that is initialized by components on startup and used across the codebase.
+In most places, it is used as a `printf`-style logger and only in some instances we make use of logrus' structured logging functionality.
+
+In the process of migrating our components to native controller-runtime components (see [gardener/gardener#4251](https://github.com/gardener/gardener/issues/4251)), we also want to make use of controller-runtime's built-in mechanisms for streamlined logging.
+controller-runtime uses [logr](https://github.com/go-logr/logr), a simple structured logging interface, for library-internal logging and logging in controllers.
+
+logr itself is only an interface and doesn't provide an implementation out of the box.
+Instead, it needs to be backed by a logging implementation like [zapr](https://github.com/go-logr/zapr). Code that uses the logr interface is thereby not tied to a specific logging implementation and makes the implementation easily exchangeable.
+controller-runtime already provides a [set of helpers](https://github.com/kubernetes-sigs/controller-runtime/tree/v0.11.0/pkg/log/zap) for constructing zapr loggers, i.e., logr loggers backed by [zap](https://github.com/uber-go/zap), which is a popular logging library in the go community.
+Hence, we are migrating our component logging from logrus to logr (backed by zap) as part of [gardener/gardener#4251](https://github.com/gardener/gardener/issues/4251).
+
+> ⚠️ `logger.Logger` (logrus logger) is deprecated in Gardener and shall not be used in new code – use logr loggers when writing new code!
```

Comment:
In https://github.com/gardener/gardener/blob/master/pkg/logger/logrus.go I see that only `Logger` var and `NewLogger ` are deprecated. Shouldn't we also deprecated the rest of the exported funcs in this file?

Reply:
Good point, done :)